src/main/java/net/wovenmc/woven/impl/tags/WovenTagsImpl.java

```diff
@@ -0,0 +1,157 @@
+		} else if (registry == Registry.ENTITY_TYPE) {
+			return (TagDelegate<T>) new TagDelegate<EntityType<?>>(id, EntityTypeTags::getTagGroup);
+		} else if (registry == Registry.FLUID) {
+			return (TagDelegate<T>) new TagDelegate<Fluid>(id, FluidTagsAccessor.getRequiredTags()::getGroup);
+		} else {
+			return new TagDelegate<T>(id, () -> (TagGroup<T>) tagGroupMap.get(registryKey));
+		}
+	}
+
+	@Override
+	public Identifier getIdentifier() {
+		return new Identifier("woven_tags", "reload_listener");
+	}
+
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	@Override
+	protected List<TagLoaderInfo<?>> prepare(ResourceManager manager, Profiler profiler) {
+		tagGroupMap.clear();
+		Map<? extends RegistryKey<? extends Registry<?>>, ? extends SimpleRegistry<?>> dynRegistries = ((DynamicRegistryManagerImplAccessor) (Object) registryManager)
+				.getRegistries();
+		List<TagLoaderInfo<?>> loaderList = new ArrayList<>();
+
+		for (Map.Entry<? extends RegistryKey<? extends Registry<?>>, ? extends SimpleRegistry<?>> entry : dynRegistries.entrySet()) {
+			RegistryKey<? extends Registry<?>> key = entry.getKey();
+			TagGroupLoader<?> loader = javacHasAwfulTypeInference(entry.getValue(),
+					"tags/" + getDataLocation(key.getValue()), key.getValue().toString());
+			Map<Identifier, Tag.Builder> prepareResult = loader.prepareReload(manager, Runnable::run).join();
+			loaderList.add(new TagLoaderInfo(key, loader, prepareResult));
+		}
+
+		for (Registry<?> registry : Registry.REGISTRIES) {
+			if (!hasVanillaTag(registry)) {
+				RegistryKey<? extends Registry<?>> key = registry.getKey();
+				TagGroupLoader<?> loader = new TagGroupLoader((Function<Identifier, Optional<?>>) registry::getOrEmpty,
+						"tags/" + getDataLocation(key.getValue()), key.getValue().toString());
+				Map<Identifier, Tag.Builder> prepareResult = loader.prepareReload(manager, Runnable::run).join();
+				loaderList.add(new TagLoaderInfo(key, loader, prepareResult));
+			}
+		}
+
+		return loaderList;
+	}
+
+	/*
+	 * This is basically just a hack to get around javac's awful type inference. It
+	 * won't acknowledge the {@link Identifier} version of {@code getOrEmpty}
+	 * without this for some reason.
+	 */
+	private <T> TagGroupLoader<T> javacHasAwfulTypeInference(Registry<T> registry, String dataType, String entryType) {
+		return new TagGroupLoader<T>(registry::getOrEmpty, dataType, entryType);
+	}
+
+	@Override
+	protected void apply(List<TagLoaderInfo<?>> loaders, ResourceManager manager, Profiler profiler) {
+		for (TagLoaderInfo<?> info : loaders) {
+			TagGroup<?> group = info.loader.applyReload(info.tags);
+			tagGroupMap.put(info.registryKey, group);
+		}
+	}
+
+	private boolean hasVanillaTag(Registry<?> registry) {
+		return registry == Registry.BLOCK || registry == Registry.ITEM || registry == Registry.ENTITY_TYPE
+				|| registry == Registry.FLUID;
+	}
+
+	private String getDataLocation(Identifier id) {
```

Comment:
this'd result in `tags/c/thing`, which isn't what we want I think

Reply:
This is the `Registry`'s identifier, not the tag's -- the javadoc on `WovenTags` goes over how this behaves. Could probably do with a comment on this method, though.

Reply:
ahhhhhh okay